Sample controller tests

gregory-m opened the pull request:
**What this PR does / why we need it**:
This PR adds unit tests to sample-controller.

**Special notes for your reviewer**:
This is currently based on the munnerz:sample-controller branch. Please don't merged it until #52753 is merged.

gregory-m commented:
CC @munnerz, @nikhita

sttts commented:
/approve no-issue

Leaving the detailed review to  @munnerz and @nikhita

sttts commented:
/ok-to-test

sttts commented:
verify tests are red.

gregory-m commented:
To pass verify tests this branch should be rebased on master to use new gazelle/kazel version. Since this PR based on `munnerz:sample-controller` it should be rebased first. I will-try to cherry-pick needed commits from master.

nikhita commented:
@gregory-m https://github.com/kubernetes/kubernetes/pull/52753 has been merged. Can you rebase? Thanks!

gregory-m committed:
Added unit tests to sample-controller

Added unit tests.

Changed NewController function to accept informers and not informers
factory. This make code more testable and align it with other
controllers.

gregory-m commented:
/retest

sttts commented:
@tnozicka could you take a look? This should serve as a best practice example for testing controllers.

gregory-m commented:
@nikhita @munnerz Any chances for review?

nikhita commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -88,13 +88,8 @@ type Controller struct {
-	deploymentInformer := kubeInformerFactory.Apps().V1beta2().Deployments()
-	fooInformer := sampleInformerFactory.Samplecontroller().V1alpha1().Foos()
+	deploymentInformer appsinformers.DeploymentInformer,
+	fooInformer informers.FooInformer) *Controller {
```
:+1:

nikhita commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,274 @@
+
+	client    *fake.Clientset
+	k8sClient *k8sfake.Clientset
+	// Objects to put in the store.
```
nit: newline here

nikhita commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,274 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	k8sinformers "k8s.io/client-go/informers"
```
maybe instead of `k8sinformers`, we can write it as `kubeinformers`? Would be nice to have uniform variable names. :)

nikhita commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,274 @@
+	t *testing.T
+
+	client    *fake.Clientset
+	k8sClient *k8sfake.Clientset
```
kubeclient

nikhita commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,274 @@
+}
+
+func (f *fixture) runExpectError(fooName string) {
+	f.runController(fooName, true, true)
```
I understand that https://github.com/kubernetes/kubernetes/blob/5adfb24f8f25a0d57eb9a7b158db46f9f46f0d80/pkg/controller/deployment/deployment_controller_test.go also uses a similar pattern but wondering why we can't use just one `runController` function and pass the parameters to it directly?

gregory-m replied:
I just think `run/runExpectError` is more self-describing, and allow to understand test code faster.

nikhita replied:
:+1:

nikhita commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,274 @@
+		return ""
+	}
+	return key
+
```
nit: extra newline

nikhita commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,274 @@
+	deploymentLister []*apps.Deployment
+
+	// Actions expected to happen on the client.
+	k8sActions []core.Action
```
kubeactions

nikhita commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,274 @@
+	actions    []core.Action
+
+	// Objects from here preloaded into NewSimpleFake.
+	k8sObjects []runtime.Object
```
kubeobjects

nikhita reviewed (commented):
@gregory-m Sorry for getting to it late! Looks great!! :tada: 

Just a few minor comments.

nikhita commented:
Looks good after a squash + fixing the conflicts. @sttts?

munnerz commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,271 @@
+	f.expectUpdateFooStatusAction(foo)
+	f.expectUpdateDeploymentAction(d)
+	f.run(getKey(foo, t))
+}
```
From what I can see, this test doesn't actually verify that the deployment has been updated to have 2 replicas, only that the deployment was updated in *some way*. Is that desired?

gregory-m replied:
This enough for internal k8s controllers, but let me check if we can do better job here.

gregory-m replied:
Ok, I figured out the better way, we can cast action objects and call .GetObject() to receive actual object and compare it with the expected one.

Only one problem, I'm on  vacation next week and will not have time to work on it.

So we have two options here:
1. Merge it as it is, and I will submit better implementation in one week
2. Wait one week for it.

gregory-m replied:
Done

gregory-m commented:
@munnerz can please take another look.

ScorpioCPH commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,310 @@
+
+func TestDoNothing(t *testing.T) {
+	f := newFixture(t)
+	foo := newFoo("test", int32Ptr(1))
```
nit: should we use different `Foo` name to avoid unexpected result?

gregory-m replied:
You mean `"test"` string passed to `newFoo`?

ScorpioCPH replied:
Yes.

ScorpioCPH reviewed (commented):
Thanks @gregory-m. Looks good to me.

nikhita commented:
@gregory-m Can you rebase?

@munnerz PTAL.

munnerz commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,310 @@
+	}
+}
+
+func filterInformerActions(actions []core.Action) []core.Action {
```
Adding a comment to explain why this is necessary would be great. I think this could catch people out if writing their own tests.

munnerz commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,310 @@
+	f.fooLister = append(f.fooLister, foo)
+	f.objects = append(f.objects, foo)
+
+	d := newDeployment(foo)
```
Can we rename this var to make it clear this is the expectedDeployment? Also a comment may be helpful

munnerz commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,310 @@
+
+		expectedAction := f.actions[i]
+		checkAction(expectedAction, action, f.t)
+	}
```
Add a note saying we check `f.actions[i]` against `actions[i]`, as f.actions is an **ordered list** of expected actions. Perhaps that note should be on the fixture struct itself 😄

gregory-m replied:
I addressed all comments except this one. `fixture` already has comments for its fields.

But since you waited this comment and you familiar with code, I can assume for most people it will be even harder to understand the code.

The problem I can't came with good comment, maybe you can help?

Thanks.

munnerz commented on staging/src/k8s.io/sample-controller/controller_test.go:
```diff
@@ -0,0 +1,310 @@
+	}
+}
+
+func checkAction(expected, actual core.Action, t *testing.T) {
```
A comment explaining `checkAction`s behaviour would be great (verifies that the action is as expected, and ensures the resources attached to the actions are equal)

munnerz reviewed (changes requested):
Sorry it's taken a while to look at this again - all looks great, just a few nits regarding comments and making it clear what each part of the code is doing (I'm being extra picky as this is an example application, so others will base their own code on what we put here).

Can we also ensure that these tests are run during the verify CI jobs? I'm not certain how it works - perhaps just by the Bazel file being updated to include a `go_test` target, it will work? ¯\_(ツ)_/¯

gregory-m commented:
/retest

gregory-m commented:
/retest

nikhita commented:
/retest

nikhita commented:
xref https://github.com/kubernetes/sample-controller/issues/5

sttts commented:
Any progress here? @munnerz are there open comments?

nikhita commented:
/retest

dims commented:
/test pull-kubernetes-typecheck

munnerz commented:
Sorry for the delay here - I'm happy with this as is.

If you can rebase (seems to be deps related), I'll add lgtm 😄

gregory-m commented:
My bad. Pushed wrong branch will fix it soon.

gregory-m commented:
/retest

gregory-m commented:
@munnerz rebased and squashed.

munnerz commented:
/lgtm

nikhita commented:
🎉
